File: orion-platform-service/src/engine/MultiTargetExecutor.ts
import { StageExecutor } from './StageExecutor';
import { GrayScaleController, ExecutionBatch } from './GrayScaleController';
import { PipelineRun } from '../models/PipelineRun';
import { PipelineExecution } from './PipelineEngine';
import { PipelineStage } from '../models/Pipeline';
import { OrionError, ErrorCode } from '../errors';
import pino from 'pino';

const logger = pino({ level: process.env.LOG_LEVEL || 'info' });

export interface TargetResult {
  target: string;
  batchIndex: number;
  success: boolean;
  error?: string;
  durationMs: number;
}

export interface BatchResult {
  batchIndex: number;
  targets: string[];
  targetResults: TargetResult[];
  batchSuccess: boolean;
}

export interface MultiTargetResult {
  stageName: string;
  executionMode: string;
  totalTargets: number;
  totalBatches: number;
  batchResults: BatchResult[];
  overallSuccess: boolean;
}

export class MultiTargetExecutor {
  private grayscaleController: GrayScaleController;
  private stageExecutor: StageExecutor;

  constructor(
    grayscaleController: GrayScaleController,
    stageExecutor: StageExecutor
  ) {
    this.grayscaleController = grayscaleController;
    this.stageExecutor = stageExecutor;
  }

  async execute(
    run: PipelineRun,
    execution: PipelineExecution,
    stage: PipelineStage
  ): Promise<MultiTargetResult> {
    const batches = this.grayscaleController.splitBatches(stage);

    if (batches.length === 0) {
      throw new OrionError(
        `Stage '${stage.name}' has no targets to execute`,
        ErrorCode.VALIDATION_ERROR
      );
    }

    const mode = stage.executionMode ?? 'oneshot';
    const batchResults: BatchResult[] = [];
    let overallSuccess = true;

    logger.info(
      { runId: run.id, stage: stage.name, mode, totalBatches: batches.length, totalTargets: batches.reduce((s, b) => s + b.targets.length, 0) },
      'MultiTargetExecutor starting'
    );

    for (const batch of batches) {
      const batchStart = Date.now();
      const targetResults: TargetResult[] = [];

      const promises = batch.targets.map(async (target) => {
        return this.executeTarget(run, execution, stage, target, batch.batchIndex);
      });
      const results = await Promise.all(promises);
      targetResults.push(...results);

      const batchDuration = Date.now() - batchStart;
      const batchSuccess = targetResults.every((r) => r.success);
      if (!batchSuccess) overallSuccess = false;

      logger.info(
        { runId: run.id, stage: stage.name, batchIndex: batch.batchIndex, batchSuccess, batchDuration }
      );

      batchResults.push({
        batchIndex: batch.batchIndex,
        targets: batch.targets,
        targetResults,
        batchSuccess,
      });

      if (!batchSuccess && mode === 'grayScale') {
        logger.warn(
          { runId: run.id, stage: stage.name, failedBatch: batch.batchIndex },
          'GrayScale batch failed, stopping remaining batches'
        );
        break;
      }
    }

    logger.info(
      { runId: run.id, stage: stage.name, overallSuccess, totalBatches: batchResults.length },
      'MultiTargetExecutor complete'
    );

    return {
      stageName: stage.name,
      executionMode: mode,
      totalTargets: batches.reduce((s, b) => s + b.targets.length, 0),
      totalBatches: batches.length,
      batchResults,
      overallSuccess,
    };
  }

  private async executeTarget(
    run: PipelineRun,
    execution: PipelineExecution,
    stage: PipelineStage,
    target: string,
    batchIndex: number
  ): Promise<TargetResult> {
    const start = Date.now();
    const targetLabel = `${stage.name}[${target}]`;

    try {
      const targetedStage: PipelineStage = {
        ...stage,
        name: targetLabel,
        steps: stage.steps.map((step) => ({
          ...step,
          name: `${step.name}-${target}`,
        })),
      };

      await this.stageExecutor.executeStage(
        run.pipelineId,
        run.id,
        targetedStage,
        execution as any
      );

      return {
        target,
        batchIndex,
        success: true,
        durationMs: Date.now() - start,
      };
    } catch (err: any) {
      logger.error(
        { runId: run.id, target, batchIndex, error: err.message },
        `Target execution failed`
      );
      return {
        target,
        batchIndex,
        success: false,
        error: err.message,
        durationMs: Date.now() - start,
      };
    }
  }
}

File: orion-platform-svc-go/internal/pipeline-engine/service/multi_target_executor.go
package service

import (
	"context"
	"fmt"
	"sync"
	"time"

	"orion/platform-svc-go/internal/pipeline-engine/models"
	"orion/platform-svc-go/internal/pipeline-engine/repository"
)

const batchTargetCount = 3

// TargetResult holds the outcome of executing a stage against a single target.
type TargetResult struct {
	Target     string  `json:"target"`
	BatchIndex int     `json:"batch_index"`
	Success    bool    `json:"success"`
	Error      *string `json:"error,omitempty"`
	DurationMs int64   `json:"duration_ms"`
}

// BatchResult aggregates the results of executing a stage across a batch of targets.
type BatchResult struct {
	BatchIndex    int            `json:"batch_index"`
	Targets       []string       `json:"targets"`
	TargetResults []TargetResult `json:"target_results"`
	BatchSuccess  bool           `json:"batch_success"`
}

// MultiTargetResult aggregates the results of executing a stage across multiple targets.
type MultiTargetResult struct {
	StageName      string        `json:"stage_name"`
	ExecutionMode  string        `json:"execution_mode"`
	TotalTargets   int           `json:"total_targets"`
	TotalBatches   int           `json:"total_batches"`
	BatchResults   []BatchResult `json:"batch_results"`
	OverallSuccess bool          `json:"overall_success"`
}

// MultiTargetExecutor dispatches stage execution across multiple targets,
// respecting batching and execution modes (oneshot vs grayScale).
type MultiTargetExecutor struct {
	stageExecutor *StageExecutor
	repo          *repository.Repository
}

// NewMultiTargetExecutor creates a new MultiTargetExecutor.
func NewMultiTargetExecutor(se *StageExecutor, repo *repository.Repository) *MultiTargetExecutor {
	return &MultiTargetExecutor{
		stageExecutor: se,
		repo:          repo,
	}
}

// Execute runs a stage across multiple targets, batched and mode-aware.
//
// Targets are grouped into batches of batchTargetCount.
// - "oneshot" mode: runs all batches, continues even if one fails.
// - "grayScale" mode: stops on the first failed batch.
//
// Within each batch, targets are executed concurrently via goroutines.
func (mte *MultiTargetExecutor) Execute(ctx context.Context, stageName, stageID string, targets []string, mode string) *MultiTargetResult {
	if len(targets) == 0 {
		return &MultiTargetResult{StageName: stageName, ExecutionMode: mode, OverallSuccess: false}
	}
	if mode == "" {
		mode = "oneshot"
	}

	var batches [][]string
	for i := 0; i < len(targets); i += batchTargetCount {
		end := i + batchTargetCount
		if end > len(targets) {
			end = len(targets)
		}
		batches = append(batches, targets[i:end])
	}

	batchResults := make([]BatchResult, 0, len(batches))
	overallSuccess := true
	var mu sync.Mutex

	for bi, batch := range batches {
		targetResults := mte.executeBatch(ctx, stageName, stageID, batch, bi)
		batchSuccess := true
		for _, tr := range targetResults {
			if !tr.Success {
				batchSuccess = false
				break
			}
		}
		if !batchSuccess {
			overallSuccess = false
			if mode == "grayScale" {
				break
			}
		}
		mu.Lock()
		batchResults = append(batchResults, BatchResult{
			BatchIndex:    bi,
			Targets:       batch,
			TargetResults: targetResults,
			BatchSuccess:  batchSuccess,
		})
		mu.Unlock()
	}

	return &MultiTargetResult{
		StageName:      stageName,
		ExecutionMode:  mode,
		TotalTargets:   len(targets),
		TotalBatches:   len(batchResults),
		BatchResults:   batchResults,
		OverallSuccess: overallSuccess,
	}
}

// executeBatch runs a stage against a batch of targets concurrently.
func (mte *MultiTargetExecutor) executeBatch(ctx context.Context, stageName, stageID string, batch []string, batchIndex int) []TargetResult {
	var wg sync.WaitGroup
	targetResults := make([]TargetResult, len(batch))
	for i, target := range batch {
		wg.Add(1)
		go func(idx int, t string) {
			defer wg.Done()
			targetResults[idx] = mte.executeTarget(ctx, stageName, stageID, t, batchIndex)
		}(i, target)
	}
	wg.Wait()
	return targetResults
}

// executeTarget runs a stage against a single target.
//
// It dispatches to the StageOrchestrator to execute the stage with the target
// injected as a variable (TARGET_NAME), then records the result.
func (mte *MultiTargetExecutor) executeTarget(ctx context.Context, stageName, stageID string, target string, batchIndex int) TargetResult {
	start := time.Now()

	// Check for context cancellation before execution
	select {
	case <-ctx.Done():
		e := ctx.Err().Error()
		return TargetResult{
			Target:     target,
			BatchIndex: batchIndex,
			Success:    false,
			Error:      &e,
			DurationMs: time.Since(start).Milliseconds(),
		}
	default:
	}

	// Build target-aware execution context
	variables := map[string]string{
		"TARGET_NAME":  target,
		"TARGET_INDEX": fmt.Sprintf("%d", batchIndex),
	}

	// Load the stage to verify it exists
	_, err := mte.repo.GetStage(ctx, "", stageID)
	if err != nil {
		errStr := fmt.Sprintf("stage not found: %v", err)
		return TargetResult{
			Target:     target,
			BatchIndex: batchIndex,
			Success:    false,
			Error:      &errStr,
			DurationMs: time.Since(start).Milliseconds(),
		}
	}

	// Load tasks for this stage
	tasks, err := mte.repo.GetTasksByStage(ctx, "", stageID)
	if err != nil {
		errStr := fmt.Sprintf("failed to load tasks for stage %s: %v", stageID, err)
		return TargetResult{
			Target:     target,
			BatchIndex: batchIndex,
			Success:    false,
			Error:      &errStr,
			DurationMs: time.Since(start).Milliseconds(),
		}
	}

	// Execute each task in the stage with target variables
	taskFailed := false
	for _, task := range tasks {
		// Skip tasks that have already been marked as completed/failed
		if task.Status != models.TaskStatusPending {
			continue
		}
		_, taskResult := mte.stageExecutor.ExecuteTask(ctx, "", stageID, &task, variables)
		if !taskResult.Success {
			taskFailed = true
			break
		}
	}

	duringMs := time.Since(start).Milliseconds()
	if taskFailed {
		errStr := fmt.Sprintf("stage %s failed for target %s", stageName, target)
		return TargetResult{
			Target:     target,
			BatchIndex: batchIndex,
			Success:    false,
			Error:      &errStr,
			DurationMs: duringMs,
		}
	}

	// All tasks succeeded
	return TargetResult{
		Target:     target,
		BatchIndex: batchIndex,
		Success:    true,
		DurationMs: duringMs,
	}
}
